src/main/scala/UARTAdapter.scala

```diff
@@ -0,0 +1,156 @@
+  // iterate using div to convert clock rate to baud
+  val (txBaudCount, txBaudWrap) = Counter(txState === sTxWait && txfifo.io.enq.ready, div)
+  val (txSlackCount, txSlackWrap) = Counter(txState === sTxIdle && uart_io.txd === 0.U && txfifo.io.enq.ready, 4)
+
+  switch(txState) {
+    is(sTxIdle) {
+      when(txSlackWrap) {
+        txData  := 0.U
+        txState := sTxWait
+      }
+    }
+    is(sTxWait) {
+      when(txBaudWrap) {
+        txState := sTxData
+      }
+    }
+    is(sTxData) {
+      when (txfifo.io.enq.ready) {
+        txData := txData | (uart_io.txd << txDataIdx)
+      }
+      when(txDataWrap) {
+        txState := Mux(uart_io.txd === 1.U, sTxIdle, sTxBreak)
+      }.elsewhen(txfifo.io.enq.ready) {
+        txState := sTxWait
+      }
+    }
+    is(sTxBreak) {
+      when(uart_io.txd === 1.U && txfifo.io.enq.ready) {
+        txState := sTxIdle
+      }
+    }
+  }
+
+  txfifo.io.enq.bits  := txData
+  txfifo.io.enq.valid := txDataWrap
+
+  val sRxIdle :: sRxStart :: sRxData :: Nil = Enum(3)
+  val rxState = RegInit(sRxIdle)
+  // iterate using div to convert clock rate to baud
+  val (rxBaudCount, rxBaudWrap) = Counter(txfifo.io.enq.ready, div)
+  // iterate through bits in byte to deserialize
+  val (rxDataIdx, rxDataWrap) = Counter(rxState === sRxData && txfifo.io.enq.ready && rxBaudWrap, DATA_WIDTH)
+
+  uart_io.rxd := 1.U
+  switch(rxState) {
+    is(sRxIdle) {
+      uart_io.rxd := 1.U
+      when (rxBaudWrap && rxfifo.io.deq.valid) {
+        rxState := sRxStart
+      }
+    }
+    is(sRxStart) {
+      uart_io.rxd := 0.U
+      when(rxBaudWrap) {
+        rxState := sRxData
+      }
+    }
+    is(sRxData) {
+      uart_io.rxd := (rxfifo.io.deq.bits >> rxDataIdx)(0)
+      when(rxDataWrap && rxBaudWrap) {
+        rxState := sRxIdle
+      }
+    }
+  }
+  rxfifo.io.deq.ready := (rxState === sRxData) && rxDataWrap && rxBaudWrap && txfifo.io.enq.ready
+
+  val sim = Module(new SimUART(uartno))
+
+  sim.io.clock := clock
+  sim.io.reset := reset.asBool
+
+  sim.io.serial.out.bits := txfifo.io.deq.bits
+  sim.io.serial.out.valid := txfifo.io.deq.valid
+  txfifo.io.deq.ready := sim.io.serial.out.ready
+
+  rxfifo.io.enq.bits := sim.io.serial.in.bits
+  rxfifo.io.enq.valid := sim.io.serial.in.valid
+  sim.io.serial.in.ready := rxfifo.io.enq.ready
+}
+
+class SimUART(uartno: Int) extends BlackBox(Map("UARTNO" -> IntParam(uartno))) with HasBlackBoxResource {
+  val io = IO(new Bundle {
+    val clock = Input(Clock())
+    val reset = Input(Bool())
+
+    val serial = Flipped(new SerialIO(DATA_WIDTH))
+  })
+
+  addResource("/testchipip/vsrc/SimUART.v")
+  addResource("/testchipip/csrc/SimUART.cc")
+  addResource("/testchipip/csrc/uart.cc")
+  addResource("/testchipip/csrc/uart.h")
+}
+
+trait CanHavePeripheryUARTWithAdapter extends HasPeripheryUART { this: BaseSubsystem =>
+}
+
+trait CanHavePeripheryUARTWithAdapterImp extends HasPeripheryUARTModuleImp {
```

Comment:
One way to avoid having this extra trait (and the confusion between this trait and the base `HasPeripheryUART` trait) is to make the `connectSimUARTs` function a global function, as opposed to a method of the `CanHavePeripheryUARTWithAdapterImp`. The function could be 

```
object ConnectUARTPrinter {
  def apply[T <: HasPeripheryUARTModuleImp](top: T, p: Parameters): Unit = {
    require(p(PeripheryUARTKey).size == p(UARTAdapterKey).size)
    top.uart.zipWithIndex.foreach{ case (dut_io, i) =>
      val uart_sim = Module(new UARTAdapter(i)(p))
      uart_sim.io.uart.txd := dut_io.txd
      dut_io.rxd := uart_sim.io.uart.rxd
  }
}
```

You can then call this function in Chipyard's testharness on the dut returned by `BuildTop`.
This seems to me to be a bit more elegant than adding the additional trait, since traits typically describe additional hardware that is added to the system, but I may be missing some sublety.

Reply:
I'm not opposed to either approach. The main reason I had this as a trait was to make it similar to the block device and the serial adapter (a single trait that adds both the DUT component as well as the TestHarness component).

Reply:
Now that I think about this more, I think I prefer having the current approach, a separate trait from the UART trait that can potentially add the UARTAdapters in.